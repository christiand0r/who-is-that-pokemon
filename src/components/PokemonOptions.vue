<template>
  <div class="pokemon-options">
    <h1>
      <img src="@/assets/qssp.png" alt="Quién es ese Pokémon" />
    </h1>
    <ul class="options-list">
      <li
        v-for="{ name, id } in parsedPokemons"
        :key="id"
        :disabled="disableOptions"
        :class="classByResponse(id)"
        @click="selectedPokemon(id)"
      >
        {{ name }}
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  name: "Pokemon Options",
  props: {
    pokemons: {
      type: Array,
      required: true,
    },
    isCorrect: {
      type: Boolean,
      default: null,
    },
    pokemonId: {
      type: Number,
      default: null,
      required: true,
    },
  },
  data() {
    return {
      userSelection: null,
    };
  },
  methods: {
    selectedPokemon(id) {
      this.$emit("selectedPokemon", id);

      //Update state for evalutaing answers
      this.userSelection = id;
    },
    classByResponse(id) {
      if (this.isCorrect === null) return;

      if (this.isCorrect) {
        //If is correct the option selected for the user is correct
        return this.userSelection === id ? "correct active" : "incorrect";
      } else {
        //The option selected for the user is wrong
        return this.userSelection === id
          ? "incorrect active"
          : this.pokemonId === id //The correct option
          ? "correct"
          : "incorrect";
      }
    },
  },
  computed: {
    parsedPokemons() {
      return this.pokemons.map((pokemon) => ({
        ...pokemon,
        name: pokemon.name.toUpperCase(),
      }));
    },
    disableOptions() {
      return this.isCorrect !== null ? true : false;
    },
  },
};
</script>

<style scoped>
h1 {
  margin: 0;
  min-width: 280px;
}

.pokemon-options {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  max-width: 480px;
  width: 100%;
}

.options-list {
  list-style: none;
  padding: 0;
  margin: 0;
  width: 100%;
}

.options-list li {
  display: block;
  background-color: var(--gray);
  border-radius: var(--normal-radius);
  color: var(--dim);
  font-size: 2.5rem;
  font-weight: var(--medium);
  line-height: 1;
  margin-bottom: 1.5rem;
  padding: 1.5rem;
  position: relative;
  text-transform: uppercase;
  transition: 0.2s background-color, 0.2s color;
}

.options-list li[disabled="true"] {
  cursor: not-allowed;
  pointer-events: none;
}

.options-list .correct {
  background-color: var(--cyan);
  color: var(--british);
}
.options-list .incorrect {
  background-color: var(--palepink);
  color: var(--blood);
}

.options-list li:hover,
.options-list li *:hover {
  cursor: pointer;
}

.options-list li:hover {
  background-color: var(--outer);
  color: var(--ghost);
}

.options-list .correct:hover {
  background-color: var(--british);
  color: var(--ghost);
}

.options-list .incorrect:hover {
  background-color: var(--blood);
  color: var(--ghost);
}

.options-list .correct.active {
  background-color: var(--british);
  color: var(--ghost);
}

.options-list .incorrect.active {
  background-color: var(--blood);
  color: var(--ghost);
}
</style>
