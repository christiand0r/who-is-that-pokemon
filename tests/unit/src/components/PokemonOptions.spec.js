import { shallowMount } from "@vue/test-utils";
import { mockPokemons } from "../../mocks/mockPokemons.mock";
import PokemonOptions from "@/components/PokemonOptions.vue";

describe("Pokemon Options - Component", () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallowMount(PokemonOptions, {
      props: {
        pokemons: mockPokemons,
        pokemonId: 70,
      },
      data() {
        return {
          userSelection: null,
        };
      },
    });
  });

  test("debe ser igual que el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debemos mostrar 4 opciones de pokemon en el orden recibido", () => {
    const options = wrapper.findAll("ul li");

    expect(options).toHaveLength(4);

    //Iterate the array and for each option checks if the text is equal to the name of the pokemon to which its position corresponds

    options.forEach((opt, index) =>
      expect(opt.text()).toBe(wrapper.vm.pokemons[index].name)
    );
  });

  test("elegir una opciÃ³n debe actualizar la data", async () => {
    const options = wrapper.findAll("ul li");

    for (let i = 0; i < options.length; i++) {
      await options[i].trigger("click");

      expect(wrapper.vm.userSelection).toBe(wrapper.vm.pokemons[i].id);
    }
  });

  test('elegir una opcion debe emitir "selectedPokemon"', () => {
    const options = wrapper.findAll("ul li");

    options.forEach((opt, index) => {
      opt.trigger("click");

      //Check if value in the array of emmits is equals to id pokemon in pokemons
      expect(wrapper.emitted("selectedPokemon")[index]).toEqual([
        wrapper.vm.pokemons[index].id,
      ]);
    });

    expect(wrapper.emitted("selectedPokemon")).toHaveLength(4);
  });
});
