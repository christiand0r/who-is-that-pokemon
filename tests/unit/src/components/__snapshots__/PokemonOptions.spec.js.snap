// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Options - Component debe ser igual que el snapshot 1`] = `
<div class="pokemon-options">
  <h1><img src="" alt="Quién es ese Pokémon"></h1>
  <ul class="options-list">
    <li disabled="false" class="">bulbasaur</li>
    <li disabled="false" class="">charmander</li>
    <li disabled="false" class="">squirtle</li>
    <li disabled="false" class="">pikachu</li>
  </ul>
</div>
`;
